kolibri/core/auth/utils.py

```diff
@@ -20,3 +26,134 @@ def create_adhoc_group_for_learners(classroom, learners):
+        "start_timestamp",
+        "end_timestamp",
+        "completion_timestamp",
+        "time_spent",
+        "progress",
+        "kind",
+        "extra_fields",
+    ],
+    UserSessionLog: [
+        "channels",
+        "start_timestamp",
+        "last_interaction_timestamp",
+        "pages",
+        "device_info",
+    ],
+    MasteryLog: [
+        "mastery_criterion",
+        "start_timestamp",
+        "end_timestamp",
+        "completion_timestamp",
+        "mastery_level",
+        "complete",
+        ("summarylog_id", ContentSummaryLog),
+    ],
+    AttemptLog: [
+        "item",
+        "start_timestamp",
+        "end_timestamp",
+        "completion_timestamp",
+        "time_spent",
+        "complete",
+        "correct",
+        "hinted",
+        "answer",
+        "simple_answer",
+        "interaction_history",
+        "error",
+        ("masterylog_id", MasteryLog),
+        ("sessionlog_id", ContentSessionLog),
+    ],
+}
+
+
+def _merge_user_models(source_user, target_user):
+    for f in user_fields:
+        source_value = getattr(source_user, f, None)
+        target_value = getattr(target_user, f, None)
+        if not target_value and source_value:
+            setattr(target_user, f, source_value)
+    target_user.save()
+
+
+def merge_users(source_user, target_user):
+    """
+    Utility to merge two users. It makes no assumptions about whether
+    the users are in the same facility and does raw copies of all
+    associated user data, rather than try to do anything clever.
+    """
+    if source_user.id == target_user.id:
+        raise ValueError("Cannot merge a user with themselves")
+
+    dataset_cache.clear()
+    dataset_cache.activate()
+
+    _merge_user_models(source_user, target_user)
+
+    id_map = {}
+
+    def _merge_log_data(source_user, target_user, LogModel):
+        log_map = {}
+        id_map[LogModel] = log_map
+        new_logs = []
+        for log in LogModel.objects.filter(user=source_user):
+            data = {}
+            for f in log_fields[LogModel]:
+                if isinstance(f, tuple):
+                    field_name, model = f
+                    data[field_name] = id_map[model][getattr(log, field_name)]
+                else:
+                    data[f] = getattr(log, f)
+            new_log = LogModel(user=target_user, **data)
+            new_log.id = new_log.calculate_uuid()
+            new_log.ensure_dataset()
+            log_map[log.id] = new_log.id
+            new_logs.append(new_log)
+        LogModel.objects.bulk_create(new_logs, batch_size=750)
+
+    _merge_log_data(source_user, target_user, ContentSessionLog)
+
+    _merge_log_data(source_user, target_user, ContentSummaryLog)
+
+    _merge_log_data(source_user, target_user, UserSessionLog)
+
+    _merge_log_data(source_user, target_user, MasteryLog)
+
+    _merge_log_data(source_user, target_user, AttemptLog)
+
+    source_user.delete()
+
+    dataset_cache.deactivate()
```

Comment:
I am trying to remember why I didn't do that in the first place - I think it might have just been because we are only using this inside the sync task at the moment, and the cache is entirely thread local.